File: C/Stack/arrayStack.c
#include <stdio.h>
#include <stdlib.h>
#include "arrayStack.h"

ARRAYSTACK* arrayStack(int capacity) {
	if (0 >= capacity) return NULL;
	
	NODE* nodes = (NODE*)calloc(capacity, sizeof(NODE));
	ARRAYSTACK* as = (ARRAYSTACK*)calloc(1, sizeof(ARRAYSTACK));

	as->capacity = capacity;
	as->nodes = nodes;
	as->top = -1;
	return as;
}

void push(ARRAYSTACK* as, NODE node) {
	if (as->capacity == as->top + 1) {
		printf("용량 초과\n");
		return;
	}
	as->top++;
	as->nodes[as->top] = node;
}

NODE pop(ARRAYSTACK* as) {
	if (as->top == -1) {
		printf("더 이상 pop할 요소가 없습니다\n");
		return;
	}
	NODE node = as->nodes[as->top];
	as->top--;
	return node;
}

void destroyStack(ARRAYSTACK* as) {
	free(as->nodes);
	free(as);
	return;
}

int main() {
	STUDENT students[] = {
		{ 3, "정왕왕" },
		{ 43, "김꽊꽊" },
		{ 29, "박멍멍" }
	};
	
	NODE nodes[3] = { 0 };
	for (size_t i = 0; i < 3; i++) {
		NODE node = { &(students[i]) };
		nodes[i] = node;
	}
	ARRAYSTACK* as = arrayStack(3);

	push(as, nodes[0]);
	push(as, nodes[1]);
	push(as, nodes[2]);

	NODE test1 = pop(as);
	STUDENT* a1 = (STUDENT*)test1.data;
	printf("%s\n", a1->name);

	destroyStack(as);
}	

File: Kotlin/Stack/ArrayStack.kt
// 추상화 정의, arrayStack으로 구현하든 linkedListStack으로 구현하든
// 동일하게 사용할 수 있게 인터페이스  정의
interface Stack<T> {
    fun push(node: T ): Unit;
    fun pop(): T;
    fun popOrNull(): T?;
}

// ArrayStack의 Node로 들어갈 클래스 
class Student(private val age: Int, private val name: String) {
    fun print() = "age : $age, name : $name";
}

class ArrayStack<T>(private val capacity: Int) : Stack<T> {
    private val myArray: Array<Any?> = arrayOfNulls(capacity)
    private var top: Int = -1;

    fun getTotalNode(): Int  = top + 1;

    override fun push(node: T): Unit {
        if (capacity == getTotalNode()) return;
        top++;
        myArray[top] = node;
    }
    override fun pop(): T {
        if (top == -1)  throw ArrayIndexOutOfBoundsException(-1);
        val node = myArray[top]; //
        myArray[top--] = null;  // 메모리 누수 방지
        return node as T;
    }
    override fun popOrNull(): T? {
        if (top == -1)  return null;
        val node = myArray.get(top); //
        myArray[top--] = null;  // 메모리 누수 방지
        return node as T?;
    }
}

fun main() {
    val myStack = ArrayStack<Any>(5);
    val students = arrayOf(
        Student(3, "김갑갑"),
        Student(5, "홍길동"),
        Student(15, "박대감"),
        Student(134, "송이버섯"),
    )
    students
        .forEach {  myStack.push(it)  }

    repeat(5) {
        val result: Student? = myStack.popOrNull() as Student?;
        println("pop 결과 : ${result?.print()}")
        println("totalNOde : ${myStack.getTotalNode()}")
    }
}